openapi-model/src/main/scala/sttp/apispec/openapi/validation/OpenAPIComparator.scala

```diff
@@ -0,0 +1,277 @@
+
+    }
+  }
+
+  private def getOperationParameters(operation: Operation): List[Parameter] = {
+    operation.parameters.flatMap {
+      case Right(parameter) => Some(parameter)
+      case Left(reference)  => resolveParameterReference(serverOpenAPI, reference.$ref)
+    }
+  }
+
+  private def resolveParameterReference(openAPI: OpenAPI, ref: String): Option[Parameter] = {
+    openAPI.components match {
+      case Some(component) => component.getLocalParameter(ref)
+      case None            => None
+    }
+  }
+
+  private def checkRequestBody(
+      clientRequestBody: RequestBody,
+      serverRequestBody: RequestBody
+  ): Option[IncompatibleRequestBody] = {
+    val contentIssues = checkContent(clientRequestBody.content, serverRequestBody.content).toList
+    if (contentIssues.nonEmpty)
+      Some(IncompatibleRequestBody(contentIssues))
+    else
+      None
+  }
+
+  private def checkResponse(clientResponse: Response, serverResponse: Response): Option[IncompatibleResponse] = {
+    val contentIssue = checkContent(clientResponse.content, serverResponse.content)
+    val headerIssues = clientResponse.headers.flatMap {
+      case (clientHeaderName, Right(clientHeader)) =>
+        val serverHeader = serverResponse.headers.get(clientHeaderName)
+        serverHeader match {
+          case Some(Right(serverHeader)) =>
+            if (clientHeader.required.getOrElse(false) && !serverHeader.required.getOrElse(false)) {
+              Some(IncompatibleRequiredHeader(clientHeaderName))
+            } else {
+              checkHeader(clientHeaderName, clientHeader, serverHeader)
+            }
+          case None => Some(MissingHeader(clientHeaderName))
+          case _    => None
+        }
+      case _ => None
+    }
+
+    val issues = contentIssue.toList ++ headerIssues
+    if (issues.nonEmpty)
+      Some(IncompatibleResponse(issues))
+    else
+      None
+  }
+
+  private def checkHeader(
+      headerName: String,
+      clientHeader: Header,
+      serverHeader: Header
+  ): Option[IncompatibleHeader] = {
+    val schemaIssues = checkSchema(clientHeader.schema, serverHeader.schema)
+    val contentIssue = checkContent(clientHeader.content, serverHeader.content)
+    val isCompatibleStyle = serverHeader.style == clientHeader.style
+    val isCompatibleExplode = serverHeader.explode == clientHeader.explode
+    val isCompatibleAllowEmptyValue = serverHeader.allowEmptyValue == clientHeader.allowEmptyValue
+    val isCompatibleAllowReserved = serverHeader.allowReserved == clientHeader.allowReserved
+
+    val issues =
+      schemaIssues.toList ++
+        contentIssue.toList ++
+        (if (!isCompatibleStyle) Some(MissMatch("style")) else None).toList ++
+        (if (!isCompatibleExplode) Some(MissMatch("explode")) else None).toList ++
+        (if (!isCompatibleAllowEmptyValue) Some(MissMatch("allowEmptyValue")) else None).toList ++
+        (if (!isCompatibleAllowReserved) Some(MissMatch("allowReserved")) else None).toList
```

Comment:
Why would you use raw magic strings to discriminate these errors? Wouldn't it be better to have separate classes to represent these incompatibility issues?